cmake/options.cmake

```diff
@@ -4,6 +4,54 @@
 # This source code is licensed in accordance with the terms specified in
 # the LICENSE file found in the root directory of this source tree.
 
+# Detect osquery version
+# The OSQUERY_VERSION cache variable will be used if set or not empty
+# Otherwise detect version through git and set it in the OSQUERY_VERSION_AUTODETECTED cache variable. If detection fails, 0.0.0 will be used.
+# Append OSQUERY_VERSION_SUFFIX cache variable to the final version if specified and the version is detected from git.
+# Verify if the final version is composed by three semver components, otherwise fail.
+# Returns the final version in OSQUERY_VERSION_INTERNAL and its components in OSQUERY_VERSION_COMPONENTS
+function(detectOsqueryVersion)
+  set(OSQUERY_VERSION "" CACHE STRING "Overrides osquery version with this value")
+  set(OSQUERY_VERSION_SUFFIX "" CACHE STRING "String to append when the version is automatically detected")
+  set(OSQUERY_VERSION_AUTODETECTED "" CACHE STRING "osquery version autodetected through git. Do not manually set." FORCE)
```

Comment:
What will this be used for? Should it be set even if OSQUERY_VERSION is provided? If not, why export it?

Reply:
In this case I guess you mean OSQUERY_VERSION_AUTODETECTED?

It will show through `cmake-gui` or `cmake -LAH`, empty.
I've just kept it simple, I can eventually add the logic of showing it only when autodetecting it, and even unsetting it when setting OSQUERY_VERSION, but I didn't think that showing it anyway, empty, would've been an issue.

Reply:
I'm wondering what it does. What does it help debug?

If they don't provide `OSQUERY_VERSION`, than `OSQUERY_VERSION` and `OSQUERY_VERSION_AUTODETECTED` are the same.

If they do provide `OSQUERY_VERSION`, than `OSQUERY_VERSION_AUTODETECTED` is both unset, and unused.

Reply:
> I'm wondering what it does. What does it help debug?
> 
> If they don't provide `OSQUERY_VERSION`, than `OSQUERY_VERSION` and `OSQUERY_VERSION_AUTODETECTED` are the same.
> 
> If they do provide `OSQUERY_VERSION`, than `OSQUERY_VERSION_AUTODETECTED` is both unset, and unused.

OSQUERY_VERSION_AUTODETECTED is set with a version only if that version has been taken with git, though indeed, it will keep the **last value** if the autodetection run at least once and then you set the version with OSQUERY_VERSION, since there's no clearing of that variable.

At least for this second case it would be better for it to be empty... if you all bear with me I'll do what you suggested (showing only when it's actively set), it's quick.